### PR TITLE
Avoid collisions in cs files generated by Grpc.Tools

### LINKED ISSUE #17672: Grpc.Tools doesn't support same filename in different folders
### What version of gRPC and what language are you using?
Grpc.Tools 1.17.1
 
### What operating system (Linux, Windows, …) and version?
Windows 10
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
C# MSBuild
 
### What did you do?
One project with two protos having the same filename but in different folders.

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <Protobuf_StandardImportsPath>$(Protobuf_StandardImportsPath);..\protos</Protobuf_StandardImportsPath>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Grpc.Tools" Version="1.17.1" PrivateAssets="All" />
    <Protobuf Include="..\protos\foo.proto" Link="protos\foo.proto" />
    <Protobuf Include="..\protos\dir1\bar.proto" Link="protos\dir1\bar.proto" />
    <Protobuf Include="..\protos\dir2\bar.proto" Link="protos\dir2\bar.proto" />
  </ItemGroup>
</Project>
```
 
### What did you expect to see?
The code should compile, generating 3 cs files.
 
### What did you see instead?
 ```
MSB3105: The item "obj\Debug\netstandard2.0\Bar.cs" was specified more than once in the "Sources" parameter.  Duplicate items are not supported by the "Sources" parameter.
```
 
### Anything else we should know about your project / environment?
We have an extensive library of protos that get reused by different projects that also share some protos. Therefore a common root path is necessary. The name `config.proto` for example is used in a couple of directories.

Until now, I used the following command to generate the files:
```
%protoc_path%\protoc.exe --proto_path=%protoc_path%\..;..\protos --csharp_out=%output_path% --csharp_opt=base_namespace --grpc_out=%output_path% --plugin=protoc-gen-grpc=%protoc-grpc_path%\grpc_csharp_plugin.exe ..\protos\foo.proto ..\protos\dir1\bar.proto ..\protos\dir2\bar.proto
```

The option `base_namespace` placed the generated files in sub-directories and avoided the problem. This problem will still remains for grpc service files, since there is no option like this for `grpc_opt`. But that problem can't be circumvented so easy right now.

/cc @kkm000

### PR DESCRIPTION
**Pull request by Falco20019**
Fixes #17672
Based on the idea from #22852

CC @jtattermusch

### CODE REVIEW

**Comment by Falco20019** [comment]
@jtattermusch Your idea mostly worked great. I only did the following modifications:
* Ensure that generated paths will be created by MSBuild
* Generate correct path if `GrpcOutputDir` is empty
* Use same path for `protodep` files and unified calls
* Added documentation and tests

I tested it with my complex use case (files on different drive, relative paths not possible, ...) and it works just great.

**Comment by jtattermusch** [comment]
Thanks! 
Adhoc artifact build http://sponge/f30437b2-8244-4046-9350-dadee6a10f37 (will post the link to Grpc.Tools nuget shortly).

**Comment by jtattermusch** [comment]
CC @kkm000

**Comment by jtattermusch** [comment]
CC @JunTaoLuo  as you contributed to Grpc.Tools in the past.

File: src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets [hunk 1]
```diff
@@ -252,6 +252,7 @@
     <!-- Ensure output directories. -->
     <MakeDir Directories="%(_Protobuf_OutOfDateProto.OutputDir)" />
     <MakeDir Directories="%(_Protobuf_OutOfDateProto.GrpcOutputDir)" />
+    <MakeDir Directories="%(Protobuf_ExpectedOutputs.RelativeDir)" />
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets:255]: Interesting trick, can you elaborate a bit more on how this works and why it works reliably? (especially in cornercases).

File: src/csharp/Grpc.Tools/DepFileUtil.cs [hunk 2]
```diff
@@ -158,21 +176,45 @@ internal static class DepFileUtil
...
+        /// <summary>
+        /// Construct relative output directory with directory hash
+        /// </summary>
+        /// <param name="outputDir">Relative path to the output directory, e. g. "out"</param>
+        /// <param name="proto">Relative path to the proto item, e. g. "foo/file.proto"</param>
+        /// <returns>
+        /// Full relative path to the directory, e. g. "out/deadbeef12345678"
+        /// </returns>
+        /// <remarks>
+        /// Since a project may contain proto files with the same filename but in different
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/DepFileUtil.cs:200]: I really like the detailed explanation of why to hash the directory paths, but right now there's basically two copies of the same comment - perhaps we could update the xmldoc of GetDepFilenameForProto to just "see details on directory path hashing in GetOutputDirWithHash", instead of copy pasting.

File: src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets (hunk 1 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets:255]: I was trying to find out how we can make sure, the generated paths exist. I looked into the MSBuild code of `ProtobufCompilerOutputs` and saw, that the information about the path is already implicitly available in `PossibleOutputs` which gets assigned to `$(Protobuf_ExpectedOutputs)`. These are already prepared to be located in `%(_Protobuf_OutOfDateProto.OutputDir)` (as this will be the input to `ProtobufCompilerOutputs` a couple lines later) and contain the generated directory hash.

`%(Protobuf_ExpectedOutputs.RelativeDir)` therefore is the union of all paths that will be generated (independent of what really is part of `_Protobuf_OutOfDateProto`, but as all those paths are necessary already for the first creation, i found that ignorable also for big sets of protos).

We might be able to improve this by restrict it like `Condition = " '%(Source)' != '' and '@(Protobuf_ExpectedOutputs)' == '' "` (as done with the definition of `_Protobuf_OutOfDateProto` in `_Protobuf_GatherStaleSimple`) but I'm not too deep into MSBuild to know if this does what I think it does. As `%(Protobuf_ExpectedOutputs.Source)` is pointing to the proto of the file this expected output is related to, this might work.

File: src/csharp/Grpc.Tools/DepFileUtil.cs (hunk 2 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/DepFileUtil.cs:200]: You are right. I first had it explained separately but in the end also moved the deps file into that folder instead of having the hash as part of the file name. Will resolve it :)

File: src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets (hunk 1 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets]: I made it more explicit by introducing `_Protobuf_ExpectedGenerated` as junction of `Protobuf_ExpectedOutputs` and `_Protobuf_OutOfDateProto`. `%(_Protobuf_ExpectedGenerated.RelativeDir)` is therefore the union of all directories of expected files that have been modified.

**Comment by Falco20019** [comment]
I put some more thoughts into the dependency files. If `Protobuf_DepFilesPath` would not point to the `OutputDirectory`, the hash folders would break. Therefore, I reverted it back to `{dirhash}_{filename}.protodep` as before. Since we have no guessing there, we also have no good option to piggy-back for folder creation. And this would also only have been optional as there are no conflicts anyway.

I also made the `MakeDir` call a bit more explicit by creating `_Protobuf_ExpectedGenerated` explicitly. This allows us to only generate folder for modified files and reduces some overhead. Usually, after the clean build, the folders should already exist anyway and on first call, this should contain ALL folders.

File: src/csharp/Grpc.Tools/ProtoCompile.cs [hunk 3]
```diff
@@ -433,6 +439,18 @@ protected override string GenerateResponseFileCommands()
             return cmd.ToString();
         }
 
+        // If possible, disambiguate output dir by adding a hash of the proto file's path
+        static string MaybeEnhanceOutputDir(string outputDir, ITaskItem[] protobufs)
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/ProtoCompile.cs:443]: Currently, the directory hash will always be added by Grpc.Tools. Two places where I'm not certain:
- should we add a way of disabling this new functionality? (currently it's always on and that could perhaps cause trouble for some users. E.g. the ones that are generating the files outside of the obj/ dir and rely on the path of the created files)
- For files with no relative dir (rooted in the project root), should we still include the hash? In this case we don't need to prefix the files and their output path would stay the same as before this PR - so for simple projects, this could save some output dir cleanup (that might otherwise be necessary when upgrading to the newer Grpc.Tools)

Comment (Falco20019) [reply src/csharp/Grpc.Tools/ProtoCompile.cs:443]: I can add an option if you like. Any preferred naming for the option?

Is files in the project root really a thing? I thought there is always at least a protos folder.

Comment (jtattermusch) [reply src/csharp/Grpc.Tools/ProtoCompile.cs:443]: Let's add the option to make it possible to disable the hashed paths when needed (= which is basically fallback to the old behavior)
Ad naming: how about `Protobuf_AvoidGeneratedFileCollisions`, `Protobuf_DisambiguateGenerateFiles` or sth in that sense?
The options can be on by default (using hashes seems like a good default behavior) and one could disable it if needed.

Comment (Falco20019) [reply src/csharp/Grpc.Tools/ProtoCompile.cs:443]: Is it OK to only have it on property level or do you also want to have it on the item level? I think property level should be enough.

Comment (Falco20019) [reply src/csharp/Grpc.Tools/ProtoCompile.cs:443]: I just saw that in `CppGeneratorServices`, we already use `GetRelativeDir` to generate intermediate paths instead of hashes. `protoc` seems to generate the folders by default for C++.

Do we want to use hashes for C# or do we also want to just create the folders using the `GetRelativeDir` approach? Right now `MaybeEnhanceOutputDir` in `ProtoCompile` is always applied, also for C++. We would need to disable that based on `Generator`.

Technically, we could do it either way. But using the relative dir also in C# would make both generators more similar. And C++ would also already have the problem of the paths getting long. So having this behavior also in C# wouldn't create a gap.

Comment (Falco20019) [reply src/csharp/Grpc.Tools/ProtoCompile.cs:443]: I found another bug in the Cpp generator. `Protobuf_PrepareClean` expects to find the expected outputs based on `Protobuf` instead of `Protobuf_Compile`. But as `ProtoRoot` is missing, the paths are not resolved correctly for out-of-project protos and cleanup won't find the files.

**Comment by Falco20019** [comment]
@jtattermusch I redid most of the stuff and adjusted the logic. I thought a bit more about the problem, especially after seeing how the guessing for C++ works. I think this is a lot cleaner and better to maintain.

The new approach that I would like to discuss:
* Added `PatchedProtobuf` to `ProtoCompilerOutputs` as way to manipulate `OutputDir` and `GrpcOutputDir`
  * Added `PatchOutputDirectory` to `GeneratorService` to do the patching
  *  Will be used by C# to add the relative dir into the `OutputDir`
  * `GetPossibleOutputs` is easier for C# since it only works in the dir without doing the hashing stuff
  * This allows us to keep `ProtoCompile` close to `protoc` call and don't have special logic only for C# generator
  * `_Protobuf_CoreCompile` does not need the cryptic adjustment (`_Protobuf_ExpectedGenerated`) for adding the generated paths since `%(_Protobuf_OutOfDateProto.[Grpc]OutputDir)` works again

I also added a fix for C++ as this was broken for protos outside of the project (`ProtoRoot` was not available in `Protobuf_PrepareClean`, therefore the relative dirs where wrong). This allowed us also to use this for C#:
* Extracted target `_Protobuf_SetProtoRoot` from `_Protobuf_SelectFiles`
  * `_Protobuf_SetProtoRoot` added as dependency to `Protobuf_Clean` to fix the bug in `Protobuf_PrepareClean`

**Comment by jtattermusch** [comment]
Adhoc package build: https://sponge.corp.google.com/c5b95681-bf71-4ecc-943d-35d2fbf41ca8

**Comment by jtattermusch** [comment]
> @jtattermusch I redid most of the stuff and adjusted the logic. I thought a bit more about the problem, especially after seeing how the guessing for C++ works. I think this is a lot cleaner and better to maintain.
> 
> The new approach that I would like to discuss:
> 
> * Added `PatchedProtobuf` to `ProtoCompilerOutputs` as way to manipulate `OutputDir` and `GrpcOutputDir`
>   
>   * Added `PatchOutputDirectory` to `GeneratorService` to do the patching
>   * Will be used by C# to add the relative dir into the `OutputDir`
>   * `GetPossibleOutputs` is easier for C# since it only works in the dir without doing the hashing stuff
>   * This allows us to keep `ProtoCompile` close to `protoc` call and don't have special logic only for C# generator (`MaybeEnhanceOutputDir` would have introduced a bug for `Generator = cpp`)
>   * `_Protobuf_CoreCompile` does not need the cryptic adjustment (`_Protobuf_ExpectedGenerated`) for adding the generated paths since `%(_Protobuf_OutOfDateProto.[Grpc]OutputDir)` works again
> * Added `Protobuf_ProtoRoot` to define a `ProtoRoot` default if outside of the project
>   
>   * This allows the relative paths to be generated a lot better
>   * It keeps the old behavior to generate everything directly in `[Grpc]OutputDir` if no `ProtoRoot` was given
>   * This also means, that if no `ProtoRoot` or `Protobuf_ProtoRoot` is given, it could still create duplicates, but they are very easy and generic fixable: `<Protobuf Include="$(LIProtoPath)/**/*.proto" ProtoRoot="$(LIProtoPath)" />` which is more natural than setting `OutputDir = "$(Protobuf_OutputPath)%(RelativeDir)"`
>   * This works for local, relative and absolute paths (including other drives)
> * I kept the changes to `DepFileUtil` (separation of `GetDepFilenameForProto` into `GetOutputDirWithHash` and `GetDirectoryHash`) as it was basically a refactoring with added documentation
>   
>   * If you would like to keep it in one method for performance, I will just revert it back to master
> 
> I also added a fix for C++ as this was broken for protos outside of the project (`ProtoRoot` was not available in `Protobuf_PrepareClean`, therefore the relative dirs were calculated wrong, leading to nothing being cleared). This allowed us also to use this for C#:
> 
> * Extracted target `_Protobuf_SetProtoRoot` from `_Protobuf_SelectFiles`
>   
>   * `_Protobuf_SetProtoRoot` added as dependency to `Protobuf_Clean` to fix the bug in `Protobuf_PrepareClean`
> 
> If you are good with this idea, I would adjust the documentation for it. Didn't want to start until I get your feedback.

To be honest I'm having trouble to understand how exactly the new approach works. I'm going to add some comments to pinpoint the places in the code where I'm unsure about what exactly it does (and therefore is hard for me to tell whether it's good/bad idea to do so).

File: src/csharp/Grpc.Tools/GeneratorServices.cs [hunk 4]
```diff
@@ -55,31 +54,58 @@ protected bool GrpcOutputPossible(ITaskItem proto)
...
+            string pathStem = Path.Combine(outdir, relative);
+            protoItem.SetMetadata(Metadata.OutputDir, pathStem);
+
+            // Override outdir if GrpcOutputDir present, default to proto output.
+            string grpcdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
+            if (grpcdir != "")
+            {
+                pathStem = Path.Combine(grpcdir, relative);
+            }
+            protoItem.SetMetadata(Metadata.GrpcOutputDir, pathStem);
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/GeneratorServices.cs:89]: I feel that updating the GrpcOutputDir and OutputDir makes the logic quite hard to follow?
Who will be the changes visible to?  Does this actually modify the input item (visible to the entire rest of the build) or is this local to this task (in which case we should probably remember the values in a different way?)

File: src/csharp/Grpc.Tools/ProtoCompilerOutputs.cs [hunk 5]
```diff
@@ -68,8 +76,13 @@ public override bool Execute()
             // Get language-specific possible output. The generator expects certain
             // metadata be set on the proto item.
             var possible = new List<ITaskItem>();
+            var patched = new List<ITaskItem>();
             foreach (var proto in Protobuf)
             {
+                // This operates on a local copy and has no effect on the MSBuild instance!
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/ProtoCompilerOutputs.cs:82]: Just saw this comment.  It seems to contradict the doc comment of the method: `// Update OutputDir and GrpcOutputDir for the item and all subsequent
        // targets using this item. This should only be done if the real
        // output directories for protoc should be modified.`

If we're actually only modifying the output dir values in this instance, there should be a less confusing way of doing this (store in a different variable / metadata?)

File: src/csharp/Grpc.Tools/ProtoCompilerOutputs.cs [hunk 6]
```diff
@@ -68,8 +76,13 @@ public override bool Execute()
...
+                generator.PatchOutputDirectory(proto);
+                patched.Add(proto);
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/ProtoCompilerOutputs.cs:84]: this seems odd. You're taking items from `Protobuf` and you're modifying them (without making a defensive copy), so once done, both Protobuf and PatchedProtobuf will contain the same instances of ITaskItem, which is very confusing (and the items in Protobuf will all be patched, which is counterintuitive given the naming).

File: src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets [hunk 7]
```diff
@@ -115,27 +116,35 @@
...
+         files outside of project directory, use each .proto file's directory as the root 
+		 or Protobuf_ProtoRoot if set. -->
     <FindUnderPath Path="$(MSBuildProjectDirectory)"
                    Files="@(Protobuf->WithMetadataValue('ProtoRoot',''))">
       <Output TaskParameter="InPath" ItemName="_Protobuf_NoRootInProject"/>
       <Output TaskParameter="OutOfPath" ItemName="_Protobuf_NoRootElsewhere"/>
     </FindUnderPath>
+	
+    <ItemGroup>
+      <Protobuf>
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets:132]: how do we know that this change won't break existing users?

File: src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets [hunk 8]
```diff
@@ -115,27 +116,35 @@
           DependsOnTargets="Protobuf_Compile"
           Condition=" '$(Language)' == 'C#' " />
 
-  <Target Name="_Protobuf_SelectFiles">
+  <Target Name="_Protobuf_SetProtoRoot">
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets:119]: Can you explain the logic around Protobuf_ProtoRoot a bit more? I'm not sure I understand the purpose of it.

**Comment by jtattermusch** [comment]
Correct me if I'm mistaken, but it looks like the new approach doesn't use dir hashing for avoiding the collisions at all?
Given that this is a complete revamp of the original logic, it would have been better if there was a separate PR for it, rather than replacing the logic in this PR - can we restore the original logic in this PR and create a separate PR with the new alternative approach?
I think I like the approach with dir hashing a bit more because it seems simpler (that said I think some ideas from the new approach still apply to the dir hashing approach).

File: src/csharp/Grpc.Tools.Tests/CSharpGeneratorTest.cs [hunk 9]
```diff
@@ -84,5 +84,15 @@ public void OutputDirMetadataRecognized()
             Assert.AreEqual(1, poss.Length);
             Assert.That(poss[0], Is.EqualTo("out/Foo.cs") | Is.EqualTo("out\\Foo.cs"));
         }
+
+        [Test]
+        public void OutputDirPatched()
+        {
+            var item = Utils.MakeItem("sub/foo.proto", "OutputDir", "out");
```
[the remark this answers is not in the document]
Comment (Falco20019) [reply src/csharp/Grpc.Tools.Tests/CSharpGeneratorTest.cs:91]: I can try to find more tests, but it's basically adding the path up to the proto file to the output dir. So `sub/foo.proto` + `out` -> `out/sub/Foo.cs`

`more/sub/folders/foo.proto` and `obj/Debug/netstandard2.1` will become `obj/Debug/netstandard2.1/more/sub/folders/Foo.cs`.

File: src/csharp/Grpc.Tools/DepFileUtil.cs [hunk 10]
```diff
@@ -269,5 +301,51 @@ static int FindLineSeparator(string line)
                 return new string[0];
             }
         }
+
+        // Calculate part of proto path relative to root. Protoc is very picky
+        // about them matching exactly, so can be we. Expect root be exact prefix
+        // to proto, minus some slash normalization.
+        internal static string GetRelativeDir(string root, string proto, TaskLoggingHelper log)
```
[the remark this answers is not in the document]
Comment (Falco20019) [reply src/csharp/Grpc.Tools/DepFileUtil.cs:308]: No new logic, only moved it without changes from `GeneratorServices.cs` to `DepFileUtil.cs`. Before, it was used only in `CppGeneratorServices` and now we use it with `CsharpGeneratorServices` too. We could also keep it in `GeneratorServices` but move it from `CppGeneratorServices` to `GeneratorServices`. Might be the better option. In an earlier version, I also needed this method in `ProtoCompile`, but that's gone.

File: src/csharp/Grpc.Tools/GeneratorServices.cs (hunk 4 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/GeneratorServices.cs:89]: This will be used to update `Protobuf_Compile` for the rest of the build and it's documented in the `targets` file to make it easier to follow. This allows us to streamline the build. We could also add a separate metadata instead of overwriting it. But then, the original `OutputDir` would not be used anywhere in the `targets`.

For C++, this is skipped as protoc is generating the sub-folders according to this logic already. For C#, we would adjust the `OutputDir` to point to the relative path as C++ and other languages would do it.

```
Identity: protos/sub/folder/foo.proto
ProtoRoot: protos
OutputDir (before patch): obj/Debug/netstandard2.1
OutputDir (after patch): obj/Debug/netstandard2.1/sub/folder
```

Therefore the plugin will generate the files to `obj/Debug/netstandard2.1/sub/folder` as if the user had manually set `OutputDir = "$(Protobuf_OutputPath)%(RelativeDir)"` but this also works for protos not on the drive since `ProtoRoot` is used and not the project location. This just replicates the C++ behavior.

File: src/csharp/Grpc.Tools/ProtoCompilerOutputs.cs (hunk 5 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/ProtoCompilerOutputs.cs:82]: I might rework this comment. The `generator.PatchOutputDirectory` will patch the instance of `proto` which is local-only to avoid having to create a copied instance. `Protobuf` is only a copy based on the input and any changes done to it won't affect MSBuild.

I could store it in a different variable, but then we would either have to decide to deep-copy `proto` which decreases performance and makes no sense since it's a local-only copy already, or if we assign it to another variable but still modify `proto`, we would need to document it properly so that everyone knows that `proto` is also patched, not just the return value.

This seemed like the most obvious without side-effects.

File: src/csharp/Grpc.Tools/ProtoCompilerOutputs.cs (hunk 6 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/ProtoCompilerOutputs.cs:84]: `Protobuf` is a local copy of the MSBuild items (see above). I can create copies and assign them to a separate variable to make it more clear, but this will result in duplicating items twice (once from MSBuild into the task and again in PatchOutputDirectory).

File: src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets (hunk 8 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets:119]: `_Protobuf_SetProtoRoot` is an extract from `_Protobuf_SelectFiles`. `Protobuf_Clean` has an bug in C++ because it's missing the `ProtoRoot`. As this PR would also use `ProtoRoot` for calculating `OutputDir` in C#, this bug would also get copied over to C#. Therefore I fixed it in here instead of in a separate PR.

But we can also put it in a separate PR which blocks this one as it's also a bug by itself.

File: src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets (hunk 7 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets:132]: It's the same logic but on `Protobuf` instead of `Protobuf_Compile` as we have to patch it for all items iterated over in `Protobuf_Clean`, not just the one that should be compiled. I only also added the logic for `$(Protobuf_ProtoRoot)` in additionally. So if you prefer, I can split it into more PRs. One for fixing the C++ bug, one for adding `$(Protobuf_ProtoRoot)` and keep this one only for the `Avoid collisions` part and have it blocked by the other 2 PRs.

**Comment by Falco20019** [comment]
I think you are right with having a separate PR and reverting this. I would even go a step further and split this up into 3 more PRs:
* Fix C++ cleanup bug
* Introduce `$(Protobuf_ProtoRoot)`
* Avoid collisions in cs file generated by Grpc.Tools (new approach)

This approach does not need the hashing, as it's duplicating the logic of the C++ generator. It just put's the files in the `ObjectDir` in folders replicating the folder structure of the proto up to the `ProtoRoot`.

So if you have only one file `K:\protos\my\sub\folder\foo.proto` and you do `<Protobuf Include="K:\**\*.proto" ProtoRoot="K:\protos" />`, the files will be generated as `obj\Debug\netstandard2.1\my\sub\folder\Foo.cs` as it already was for C++. This is also the more natural approach to `protoc` as you replicate the folder structure as `protoc` is using it in the `import`s. It also generates folder names that you can rely on. This is also independent of where your csproj is located at, since the folders are relative to `ProtoRoot` and not the csproj.

**Comment by jtattermusch** [comment]
> I think you are right with having a separate PR and reverting this. I would even go a step further and split this up into 3 more PRs:
> 
> * Fix C++ cleanup bug
> * Introduce `$(Protobuf_ProtoRoot)`
> * Avoid collisions in cs file generated by Grpc.Tools (new approach)

Sg, let's split this into multiple PRs. The smaller PRs, the easier they are to integrate.

> 
> This approach does not need the hashing, as it's duplicating the logic of the C++ generator. It just put's the files in the `ObjectDir` in folders replicating the folder structure of the proto up to the `ProtoRoot`.
> 
> So if you have only one file `K:\protos\my\sub\folder\foo.proto` and you do `<Protobuf Include="K:\**\*.proto" ProtoRoot="K:\protos" />`, the files will be generated as `obj\Debug\netstandard2.1\my\sub\folder\Foo.cs` as it already was for C++. This is also the more natural approach to `protoc` as you replicate the folder structure as `protoc` is using it in the `import`s. It also generates folder names that you can rely on. This is also independent of where your csproj is located at, since the folders are relative to `ProtoRoot` and not the csproj.
> 
> So if you don't think too deep into it, this is a lot easier to understand for the user than random hashes. It just replicates the structure that he already had for his proto files.

Ok, thanks for the explanation, I'll take another look as soon as you split the PR.

**Comment by Falco20019** [comment]
Extracted the first two PRs:
* #22895
* #22896

Introducing `$(Protobuf_ProtoRoot)` is build on top of the #22895. I will create the remaining PRs once these two are fixed to not have to rebase a lot.

File: src/csharp/Grpc.Tools/DepFileUtil.cs [hunk 11]
```diff
@@ -269,5 +301,51 @@ static int FindLineSeparator(string line)
...
+                    proto, root);
+                return "";
+            }
+            return protoDir.Substring(rootDir.Length);
+        }
+
+        // './' or '.\', normalized per system.
+        internal static string s_dotSlash = "." + Path.DirectorySeparatorChar;
+
+        internal static string EndWithSlash(string str)
```
Comment (JunTaoLuo) [inline src/csharp/Grpc.Tools/DepFileUtil.cs:336]: nit: `EnsureTrailingSlash` sound more descriptive for the purpose of this method.

File: src/csharp/Grpc.Tools/DepFileUtil.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply src/csharp/Grpc.Tools/DepFileUtil.cs:308]: I think moving it to GeneratorServices seems like the better choice.

File: src/csharp/Grpc.Tools/GeneratorServices.cs [hunk 12]
```diff
@@ -55,31 +54,58 @@ protected bool GrpcOutputPossible(ITaskItem proto)
...
             outputs[0] = Path.Combine(outdir, filename) + ".cs";
 
             if (doGrpc)
             {
-                // Override outdir if kGrpcOutputDir present, default to proto output.
                 string grpcdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
                 filename = LowerUnderscoreToUpperCamelGrpcWay(basename);
-                outputs[1] = Path.Combine(
-                    grpcdir != "" ? grpcdir : outdir, filename) + "Grpc.cs";
+                outputs[1] = Path.Combine(grpcdir, filename) + "Grpc.cs";
```
Comment (JunTaoLuo) [inline src/csharp/Grpc.Tools/GeneratorServices.cs:108]: Other than this line, are the changes in this method before this line needed? Also, this seems like a breaking change, how will users relying on existing behaviour be affected?

File: src/csharp/Grpc.Tools/DepFileUtil.cs (hunk 11 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/DepFileUtil.cs:336]: Only moved it from GeneratorServices. Will keep it there and rename it.

File: src/csharp/Grpc.Tools/GeneratorServices.cs (hunk 12 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/GeneratorServices.cs:108]: Grpcdir will be set in PatchOutputDir. So the behavior stays the same in this case. Only real behavior change is that it uses relative paths now based on what is set for ProtoRoot. If ProtoRoot is not set specifically, it will be set in MSBuild to the relative path of the file and that will result in the old behavior.

**Comment by Falco20019** [comment]
@JunTaoLuo I will clean this PR up once the other two PRs above are merged. Will also move stuff back to ServiceGenerator as it’s the better location to keep it there.

**Comment by jtattermusch** [comment]
> @JunTaoLuo I will clean this PR up once the other two PRs above are merged. Will also move stuff back to ServiceGenerator as it’s the better location to keep it there.

@Falco20019  Both PRs are merged now. Can you please rebase / cleanup this PR?

**Comment by Falco20019** [comment]
> @Falco20019 Both PRs are merged now. Can you please rebase / cleanup this PR?

Thanks, will look into it next week.

File: src/csharp/Grpc.Tools/GeneratorServices.cs (hunk 12 above)
Comment (Falco20019) [reply src/csharp/Grpc.Tools/GeneratorServices.cs:155]: I made some changes to keep the diff minimal as requested.

**Comment by Falco20019** [comment]
@jtattermusch @JunTaoLuo 
I rebased and cleaned this up so that only the relevant parts remained and that the changes are easier to follow.

* Added `PatchedProtobuf` to `ProtoCompilerOutputs` as way to manipulate `OutputDir` and `GrpcOutputDir`
  * Added `PatchOutputDirectory` to `GeneratorService` to do the patching
  *  Will be used by C# to add the relative dir into the `OutputDir`
  * `GetPossibleOutputs` is easier for C# since it only works in the dir without doing the hashing stuff
  * This allows us to keep `ProtoCompile` close to `protoc` call and don't have special logic only for C# generator
* Added `Protobuf_ProtoRoot` to define a `ProtoRoot` default if outside of the project
  * This allows the relative paths to be generated a lot better
  * It keeps the old behavior to generate everything directly in `[Grpc]OutputDir` if no `ProtoRoot` was given
  * This also means, that if no `ProtoRoot` or `Protobuf_ProtoRoot` is given, it could still create duplicates, but they are very easy and generic fixable: `<Protobuf Include="$(LIProtoPath)/**/*.proto" ProtoRoot="$(LIProtoPath)" />` which is more natural than setting `OutputDir = "$(Protobuf_OutputPath)%(RelativeDir)"`
  * This works for local, relative and absolute paths (including other drives)

File: src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets [hunk 13]
```diff
@@ -135,6 +135,7 @@
       <!-- Out-of-project files will have respective ProtoRoot='%(RelativeDir)'. -->
       <Protobuf_Rooted Include="@(_Protobuf_NoRootElsewhere)">
         <ProtoRoot>%(RelativeDir)</ProtoRoot>
+		<ProtoRoot Condition=" '$(Protobuf_ProtoRoot)' != '' ">$(Protobuf_ProtoRoot)</ProtoRoot>
```
[the remark this answers is not in the document]
Comment (Falco20019) [reply src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets]: Thanks, my IDE's settings messed it up. Fixed it!

**Comment by jtattermusch** [comment]
adhoc artifact build: https://sponge.corp.google.com/invocation?tab=Build+Log&id=3cb0cb47-36d1-4a9e-a67d-f488e3107f49&searchFor=

**Comment by jtattermusch** [comment]
Adhoc build of nugets is here: https://storage.googleapis.com/grpc-testing-kokoro-prod/test_result_public/prod/grpc/core/master/windows/grpc_build_packages/12602/20200623-064408/github/grpc/artifacts/csharp_nugets_windows_dotnetcli.zip

**Comment by jtattermusch** [comment]
@apolcyn  please add another Googler's LGTM.
